ui/src/app/kvm/views/KVMList/AddKVMForm/AddKVMForm.tsx

```diff
@@ -0,0 +1,162 @@
+} from "app/base/selectors";
+import {
+  useAddMessage,
+  useAllPowerParameters,
+  usePowerParametersSchema,
+  useWindowTitle,
+} from "app/base/hooks";
+import { PowerType, TSFixMe } from "app/base/types";
+import { formatPowerParameters } from "app/utils";
+import AddKVMFormFields from "./AddKVMFormFields";
+import FormCard from "app/base/components/FormCard";
+import FormikForm from "app/base/components/FormikForm";
+import FormCardButtons from "app/base/components/FormCardButtons";
+
+const generateAddKVMSchema = (parametersSchema: TSFixMe) =>
+  Yup.object().shape({
+    name: Yup.string(),
+    pool: Yup.string().required("Resource pool required"),
+    power_parameters: Yup.object().shape(parametersSchema),
+    type: Yup.string().required("KVM host type required"),
+    zone: Yup.string().required("Zone required"),
+  });
+
+export type AddKVMFormValues = { [x: string]: TSFixMe };
+
+export const AddKVMForm = (): JSX.Element => {
+  const dispatch = useDispatch();
+
+  const podSaved = useSelector(podSelectors.saved);
+  const podSaving = useSelector(podSelectors.saving);
+  const podErrors = useSelector(podSelectors.errors);
+  const powerTypes = useSelector(generalSelectors.powerTypes.get);
+  const powerTypesLoaded = useSelector(generalSelectors.powerTypes.loaded);
+  const resourcePools = useSelector(resourcePoolSelectors.all);
+  const resourcePoolsLoaded = useSelector(resourcePoolSelectors.loaded);
+  const zones = useSelector(zoneSelectors.all);
+  const zonesLoaded = useSelector(zoneSelectors.loaded);
+
+  const [hostType, setHostType] = useState("");
+  const [savingPod, setSavingPod] = useState(false);
+
+  const allLoaded = powerTypesLoaded && resourcePoolsLoaded && zonesLoaded;
+  const initialHostType = "virsh";
+
+  // Fetch all data required for the form.
+  useEffect(() => {
+    dispatch(generalActions.fetchPowerTypes());
+    dispatch(resourcePoolActions.fetch());
+    dispatch(zoneActions.fetch());
+  }, [dispatch]);
+
+  // Set host type in component state once power types have loaded, in order to
+  // be able to correctly format power parameters on form submit.
+  useEffect(() => {
+    const initialHostPowerType = powerTypes.find(
+      (type: PowerType) => type.name === initialHostType
+    );
+    if (initialHostPowerType && !hostType) {
+      setHostType(initialHostPowerType);
+    }
+  }, [hostType, powerTypes]);
+
+  useWindowTitle("Add KVM");
+
+  useAddMessage(
+    podSaved,
+    podActions.cleanup,
+    `${savingPod} added successfully.`,
+    setSavingPod
+  );
+
+  const AddKVMSchema = usePowerParametersSchema(
+    hostType,
+    generateAddKVMSchema,
+    true
+  );
+
+  const allPowerParameters = useAllPowerParameters(powerTypes);
+
+  let errors = "";
+  if (podErrors && typeof podErrors === "string") {
+    errors = podErrors;
+  } else if (podErrors && typeof podErrors === "object") {
+    Object.keys(podErrors).forEach((key) => {
+      errors = errors + `${podErrors[key]} `;
+    });
+  }
```

Comment:
This makes me sad.

Reply:
It might help to turn this into a hook/util and have it hidden away